File: server/src/main/java/odm/clarity/woleh/places/util/PlaceNameNormalizer.java
package odm.clarity.woleh.places.util;

import java.text.Normalizer;
import java.util.Locale;

import odm.clarity.woleh.common.error.PlaceNameValidationException;

import org.springframework.stereotype.Component;

/**
 * Normalizes place names for equality checks and matching.
 *
 * Pipeline (applied in order per PLACE_NAMES.md §1):
 *   1. Trim          — strip leading/trailing Unicode whitespace
 *   2. NFC           — Unicode Normalization Form C (composed accents)
 *   3. Case fold     — toLowerCase(Locale.ROOT); language-neutral for v1 Ghana-English
 *   4. Collapse      — maximal internal whitespace runs → single ASCII space; re-trim
 *
 * The result is used ONLY for comparison and deduplication.
 * The user-entered display string is stored separately.
 */
@Component
public class PlaceNameNormalizer {

	/** Maximum allowed Unicode code points (scalar values) per place name. */
	public static final int MAX_CODE_POINTS = 200;

	/**
	 * Normalize {@code input} through the four-step pipeline.
	 * Returns an empty string for null or blank input (callers should reject
	 * empty normalized names via {@link #validatePlaceName}).
	 */
	public String normalize(String input) {
		if (input == null) {
			return "";
		}

		// Step 1: trim (strip is Unicode-aware in Java 11+)
		String s = input.strip();

		// Step 2: NFC — composed form ensures e.g. decomposed accents equal precomposed ones
		s = Normalizer.normalize(s, Normalizer.Form.NFC);

		// Step 3: case fold (Locale.ROOT = language-neutral)
		s = s.toLowerCase(Locale.ROOT);

		// Step 4: collapse internal whitespace; re-strip in case platform adds edge spaces
		s = s.replaceAll("\\s+", " ").strip();

		return s;
	}

	/**
	 * Validate a raw (user-entered) place name before storing.
	 * Throws {@link PlaceNameValidationException} if the name is invalid.
	 *
	 * Rules:
	 * <ul>
	 *   <li>Non-empty after trim</li>
	 *   <li>At most {@value #MAX_CODE_POINTS} Unicode code points (scalar values)</li>
	 * </ul>
	 */
	public void validatePlaceName(String raw) {
		if (raw == null || raw.strip().isEmpty()) {
			throw new PlaceNameValidationException("Place name must not be empty");
		}
		int codePoints = raw.codePointCount(0, raw.length());
		if (codePoints > MAX_CODE_POINTS) {
			throw new PlaceNameValidationException(
					"Place name must not exceed " + MAX_CODE_POINTS + " Unicode code points");
		}
	}
}

File: mobile/lib/core/place_name_normalizer.dart
// ignore_for_file: avoid_classes_with_only_static_members

import 'package:unorm_dart/unorm_dart.dart' as unorm;

/// Maximum number of Unicode code points a raw place name may contain.
/// Must match the server constant in {@code PlaceNameNormalizer.MAX_CODE_POINTS}.
const int maxPlaceNameCodePoints = 200;

/// Normalizes a raw place name for **equality comparison and matching**
/// (PLACE_NAMES.md §1 — v1 pipeline).
///
/// Pipeline applied in order:
/// 1. **Trim** — strip leading/trailing Unicode whitespace.
/// 2. **NFC** — compose to Unicode Normalization Form C via [unorm_dart].
///    Ensures that logically identical text from different keyboards (e.g.
///    precomposed vs decomposed accents) compares equal.
/// 3. **Case fold** — [String.toLowerCase] (sufficient for Latin-script
///    Ghana-English place names in v1; does not cover Turkish İ/ı or other
///    locale-specific case pairs — document and revisit if needed).
/// 4. **Collapse internal whitespace** — replace every maximal run of
///    whitespace with a single ASCII space; trim again.
///
/// Returns an empty string for blank-only or empty input.
/// The result is safe to use only for **equality checks and set operations**;
/// do not display it as a user-facing string.
String normalizePlaceName(String input) {
  var s = input.trim(); // step 1: trim
  s = unorm.nfc(s); // step 2: NFC
  s = s.toLowerCase(); // step 3: case fold
  s = s.replaceAll(RegExp(r'\s+'), ' ').trim(); // step 4: collapse whitespace
  return s;
}

/// Validates a raw place name and returns an error message, or `null` if valid.
///
/// Rules (mirror of server `PlaceNameNormalizer.validatePlaceName`):
/// - Must be non-null and non-empty after trim.
/// - Must not exceed [maxPlaceNameCodePoints] Unicode code points.
///
/// Intended for use in form field validators:
/// ```dart
/// validator: (v) => validatePlaceName(v),
/// ```
String? validatePlaceName(String? raw) {
  if (raw == null || raw.trim().isEmpty) {
    return 'Place name must not be empty';
  }
  final codePointCount = raw.runes.length;
  if (codePointCount > maxPlaceNameCodePoints) {
    return 'Place name must not exceed $maxPlaceNameCodePoints Unicode code points';
  }
  return null;
}
